src/tbp/monty/frameworks/config_utils/config_args.py

```diff
@@ -1153,6 +1154,189 @@ class FiveLMMontySOTAConfig(FiveLMMontyConfig):
+                tolerances={
+                    "patch": {
+                        "hsv": np.array([0.1, 0.2, 0.2]),
+                        "principal_curvatures_log": np.ones(2),
+                    }
+                },
+                feature_weights={
+                    "patch": {
+                        "hsv": np.array([1, 0.5, 0.5]),
+                    }
+                }
+            }
+            ```
+          When we are constructing the config for learning module **i**, that entry
+          for `"patch"` will be replaced with `"patch_{i}"`. This is currently
+          done for three possible items in `learning_module_args`:
+          `"graph_delta_thresholds"`, `"tolerances"`, and `"feature_weights"`.
+
+
+    Args:
+        n_lms (int): Number of learning modules.
+        monty_class (type): Monty class.
+        learning_module_class (type): Learning module class.
+        learning_module_args (dict, optional): Arguments for learning modules.
+        sensor_module_class (type): Sensor module class.
+        sensor_module_args (dict, optional): Arguments for sensor modules.
+        motor_system_class (type): Motor system class.
+        motor_system_args (Mapping, optional): Arguments for motor system.
+        monty_args (Mapping, MontyArgs, optional): Arguments for monty.
+
+    Returns:
+        `MontyConfig`: complete monty config for multi-LM experiment.
+    """
+    # Make learning module configs.
+    if learning_module_args is None:
+        learning_module_args = {}
+    learning_module_configs = {}
+    for i in range(n_lms):
+        lm_args_i = copy.deepcopy(learning_module_args)
+        # Rename specs keyed with "patch" to "patch_{i}".
+        for name in ["graph_delta_thresholds", "tolerances", "feature_weights"]:
+            if name in lm_args_i:
+                spec = lm_args_i[name]
+                if "patch" in spec:
+                    spec[f"patch_{i}"] = spec.pop("patch")
+        learning_module_configs[f"learning_module_{i}"] = {
+            "learning_module_class": learning_module_class,
+            "learning_module_args": lm_args_i,
+        }
+
+    # Make sensor module configs.
+    if sensor_module_args is None:
+        sensor_module_args = {}
+    sensor_module_configs = {}
+    for i in range(n_lms):
+        sm_args_i = copy.deepcopy(sensor_module_args)
+        sm_args_i["sensor_module_id"] = f"patch_{i}"
+        sensor_module_configs[f"sensor_module_{i}"] = {
+            "sensor_module_class": sensor_module_class,
+            "sensor_module_args": sm_args_i,
+        }
+    sensor_module_configs["view_finder"] = {
+        "sensor_module_class": DetailedLoggingSM,
+        "sensor_module_args": {
+            "sensor_module_id": "view_finder",
+            "save_raw_obs": True,
+        },
+    }
+
+    # Make motor system config.
+    if motor_system_args is None:
+        motor_system_args = {}
+    else:
+        motor_system_args = copy.deepcopy(motor_system_args)
```

Comment:
is this line needed? What does it do?

Reply:
If we have an empty dictionary in the function signature (i., `motor_system_args={}`), the empty dictionary can be modified, and those changes will persist for future function calls. That's why I use `None` in the signature and replace it with an empty dict within the function ([example](https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python)). Is that what you're referring to?

Reply:
I understand why you have 
```
    if motor_system_args is None:
        motor_system_args = {}
```
my question was just whether `motor_system_args = copy.deepcopy(motor_system_args)` is necessary. I assume its so that if `motor_system_args` is use somewhere else the changes we make here won't affect it? I'm just wondering in what  context this would happen?

Reply:
Ah OK, sorry, I misunderstood. I thought it would probably be safer to copy any configs passed in so modifying `motor_system_args` after an experiment dict has been created wouldn't unexpectedly affect the experiment. For example, this could happen if, after you've defined one experiment config, you change something in a `motor_system_config`for use in the next experiment config (e.g., `NaiveScanSpiral`'s `step_size`), and it ends up affecting the previously defined experiment config. To me that would violate my implicit expectation that code lower in a file doesn't affect code higher up in the file. That said, a situation like that seems like not a great usage pattern, and I'd probably expect (hope?) someone would manually copy and configs that will be changed and used in next experiment config as in benchmark experiment.

I guess the other argument for copying would be to be consistent with the copying of sensor and learning module config args. It's not really a 1:1 analogy though since SM and LM configs actually have to be copied, whereas we only have one agent/motor system.

In general, I suppose I've been compartmentalizing individual experiment configs and protect them from being altered in ways I wouldn't expect, and I think that tendency just expressed itself here. That might just be a personal style thing though. Once I define an experiment config, I expect that if I want to change anything within it, I'll have to reach into the experiment config rather than changing an object that was used to initialize the experiment config.

I'm glad you brought this up though since it's revealing some unconscious patterns I've been using. Happy to go either way with this kind of stuff.

Reply:
That all makes sense, lets leave it how you have it